go/beacon/api/api.go

```diff
@@ -15,41 +16,155 @@ const (
 
 	// BeaconSize is the size of the beacon in bytes.
 	BeaconSize = 32
+
+	// EpochInvalid is the placeholder invalid epoch.
+	EpochInvalid EpochTime = 0xffffffffffffffff // ~50 quadrillion years away.
+
+	// BackendInsecure is the name of the insecure backend.
+	BackendInsecure = "insecure"
+
+	// BackendSCRAPE is the name of the SCRAPE backend.
+	BackendSCRAPE = "scrape"
 )
 
 // ErrBeaconNotAvailable is the error returned when a beacon is not
 // available for the requested height for any reason.
 var ErrBeaconNotAvailable = errors.New(ModuleName, 1, "beacon: random beacon not available")
 
-// Backend is a random beacon implementation.
+// EpochTime is the number of intervals (epochs) since a fixed instant
+// in time/block height (epoch date/height).
+type EpochTime uint64
+
+// Backend is a random beacon/time keeping implementation.
 type Backend interface {
+	// GetBaseEpoch returns the base epoch.
+	GetBaseEpoch(context.Context) (EpochTime, error)
+
+	// GetEpoch returns the epoch number at the specified block height.
+	// Calling this method with height `consensus.HeightLatest`, should
+	// return the epoch of latest known block.
+	GetEpoch(context.Context, int64) (EpochTime, error)
+
+	// GetEpochBlock returns the block height at the start of the said
+	// epoch.
+	GetEpochBlock(context.Context, EpochTime) (int64, error)
+
+	// WatchEpochs returns a channel that produces a stream of messages
+	// on epoch transitions.
+	//
+	// Upon subscription the current epoch is sent immediately.
+	WatchEpochs() (<-chan EpochTime, *pubsub.Subscription)
+
+	// WatchLatestEpoch returns a channel that produces a stream of
+	// messages on epoch transitions. If an epoch transition happens
+	// before the previous epoch is read from the channel, the old
+	// epochs are overwritten.
+	//
+	// Upon subscription the current epoch is sent immediately.
+	WatchLatestEpoch() (<-chan EpochTime, *pubsub.Subscription)
+
 	// GetBeacon gets the beacon for the provided block height.
-	// Calling this method with height `0`, should return the
-	// beacon for latest finalized block.
+	// Calling this method with height `consensus.HeightLatest` should
+	// return the beacon for the latest finalized block.
 	GetBeacon(context.Context, int64) ([]byte, error)
 
 	// StateToGenesis returns the genesis state at specified block height.
 	StateToGenesis(context.Context, int64) (*Genesis, error)
```

Comment:
Can we also include `ConsensusParameters(context.Context, int64) (*ConsensusParameters, error)` here, to make it consistent with other backends. #3565 added this method to some backends that were missing it (e.g. Scheduler), but for Epochtime/Beacon it was deferred for this PR (mostly because Epochtime was missing a grpc service, so it didn't made sense temporary adding it, as this PR would obsolete it anyway).

#3565 also added `EpochtimeConsensusParameters` to the consensus backend, which can be removed/replaced with the ConsesusParameters method on the new beacon backend.

Reply:
Was it really necessary to add the EpochtimeConsensusParameters stuff while I still have this WIP branch that massively refactors everything.

Reply:
> Was it really necessary

yeah before there was no way to query consensus parameters for some backends without dumping the whole state. And one of the tests needed the epochtime consensus parameters.